### ISSUE TITLE
postgres query takes multiple times longer than in qgis 2.18 or when run using other software

### ISSUE THREAD

**Issue by beerockxs** [qgis/QGIS#38256]
I have a few views using generated geometry that take a few seconds to execute when ran using QGIS 2.18 or in other software, such as pgAdmin 4. When run using QGIS 3.15, they take ~3.5 times as  long, a query that executes in 3.7 seconds in QGIS 2.18 takes about 13 seconds in QGIS 3.15

QGIS version | 3.15.0-Master | QGIS code revision | 48619cf552
-- | -- | -- | --
Compiled against Qt | 5.11.2 | Running against Qt | 5.11.2
Compiled against GDAL/OGR | 3.2.0dev | Running against GDAL/OGR | 3.2.0dev
Compiled against GEOS | 3.8.1-CAPI-1.13.3 | Running against GEOS | 3.8.1-CAPI-1.13.3
Compiled against SQLite | 3.29.0 | Running against SQLite | 3.29.0
PostgreSQL Client Version | 11.5 | SpatiaLite Version | 4.3.0
QWT Version | 6.1.3 | QScintilla2 Version | 2.10.8
Compiled against PROJ | 7.2.0 | Running against PROJ | Rel. 7.2.0, November 1st, 2020
OS Version | Windows 7 SP 1 (6.1) | This copy of QGIS writes debugging output.
Active python plugins | SpreadsheetLayers; db_manager; MetaSearch; processing

**Comment by gioman** [comment]
@beerockxs is this about the time that a query takes to run in the SQL panel of QGIS DB Manager?

**Comment by gioman** [comment]
> @beerockxs is this about the time that a query takes to run in the SQL panel of QGIS DB Manager?

@beerockxs is yes, please see/comment here: https://lists.osgeo.org/pipermail/qgis-developer/2020-July/061797.html

**Comment by beerockxs** [comment]
Yes, it's from the SQL Panel in QGIS DB Manager. The very same query takes ca. 0.05 seconds in QGIS 3.10, and ca. 1-1.3 seconds in QGIS 3.12 or 3.15.

**Comment by gioman** [comment]
> Yes, it's from the SQL Panel in QGIS DB Manager. The very same query takes ca. 0.05 seconds in QGIS 3.10, and ca. 1-1.3 seconds in QGIS 3.12 or 3.15.

@beerockxs thanks, I also noticed it but then I was thinking about a local issue. Great that is confirmed.

**Comment by beerockxs** [comment]
Adding "LIMIT 50000" to the query as suggested in the mailing list does not affect the speed in any way. I also do not have a huge databast with thousands of tables.

**Comment by elpaso** [comment]
First: QGIS is a GIS-centered application, we make a lot of additional controls and it is expected that QGIS will never be as fast as dedicated DB-centered application can be.

That said, there have been some recent patches that might have improved the situation, can you please test one of the nightly builds and report if the situation has improved?

**Comment by gioman** [comment]
> That said, there have been some recent patches that might have improved the situation, can you please test one of the nightly builds and report if the situation has improved?

@elpaso I used DB Manager recently to run a set of queries, albeit on some small dataset (and only on 3.10, not sure the patches you refer have been backported). It worked fine, but still seemed and checked to be slower than other backend. That said, if we have to live with it because of the additional controls I think that then we should close this.

**Comment by elpaso** [comment]
@gioman the patch I'm talking about is https://github.com/qgis/QGIS/pull/38870

**Comment by gioman** [comment]
> @gioman the patch I'm talking about is #38870

@elpaso so I haven't used a patched version. I will give it a go.

**Comment by beerockxs** [comment]
Using current Master (rev 3119eb7de2), queries are still ca. 4 times as slow as in QGIS 3.10, e.g. 4 seconds in 3.10, 16 seconds in current Master.

**Comment by Jeroen-GroeneBij** [comment]
As author of 40463 I am sorry I missed this issue. It seems to be the same.

@elpaso I understand qgis is a gis centered application. However, DB Manager is part of the core. So imho a significantly slower operating db manager in newer qgis versions is something we should try to avoid. Unfortunately, other than testing I have no other means (knowledge) to help :)

**Comment by FERRATON** [comment]
I am adding my remarks in this ticket in the hope that they may be useful 
A simple query like
SELECT * FROM mytable
is much slower in execution with QGIS 3.16 or 3.20 than with QGIS 3.10
I suspect a problem in connector.py but I didn't get into the code enough.
I see for example that the function _execute (line 98) is executed twice, but maybe this is normal?

I made comparisons with the different possibilities of executing SQL queries

QGIS 3.21 DBManager : 178 s
QGIS 3.21 execute SQL core implementation : 132s
QGIS 3.21 PostgreSQL algorithm execute SQL : 98 s
QGIS 3.10 DBManager : 127s
QGIS 3.10 algorithm PostgreSQL execute SQL : 102 s

Execute SQL core implemtation seems faster than DbManager, but less than the algorithm.

**Comment by FERRATON** [comment]
if we use the time python library
we can show that it is the call
self.result = self._toStrResultSet(self.connection.executeSql(self.sql, feedback=self.feedback)) in connector.py (QGIS 3.16)
which is much more time consuming than
cursor.execute(sql) (QGIS 3.10)

**Comment by nyalldawson** [comment]
As per https://github.com/qgis/QGIS-Enhancement-Proposals/blob/master/qep-426-demote_dbmanager.md, the db manager plugin is no longer maintained by the QGIS project. Closing as a wont-fix.
